urllib3/util/hpkp.py

```diff
@@ -0,0 +1,239 @@
+
+        This checks whether there are any ``KnownPinnedHost``s for the given
+        connection. If they are, validates that the presented certificate
+        actually matches the pin.
+        """
+        hosts = self.db.get_hosts_by_domain(domain)
+        if not hosts:
+            # This connection has not previously done any key pinning, so
+            # there's no validation to do.
+            return
+
+        host = self._find_valid_pinned_host(domain, hosts)
+        if host is None:
+            # Again, no previous pinning was done on this connection, so
+            # there's no validation to do.
+            return
+
+        # If we got this far, we have a KnownPinnedHost in hand that applies to
+        # this connection. Get the certificate and check it matches one of the
+        # pins.
+        # TODO: Actually do this!
+        return
+
+    def process_response(self, domain, response):
+        """
+        Processes a HTTPS response for HPKP.
+
+        This method checks whether the PKP header is present on the request,
+        and if it is updates its store of key pins. It also validates the
+        connection against the HPKP header, confirming that it does pass at
+        this stage.
+        """
+        # TODO: Actually do this!
+        # Some notes here: This should handle the case that the domain is an
+        # IP address, where we should not add a pin.
+        pass
+
+    def _find_valid_pinned_host(self, domain, hosts):
+        """
+        Locates the most specific valid ``KnownPinnedHost`` for a given domain.
+
+        If none of the owned hosts are valid for this domain, returns ``None``.
+        """
+        for host in hosts:
+            # First, check whether this host is still valid. If it's not, throw
+            # it away and move on.
+            if (host.start_date + host.max_age) > time.time():
+                self.db.invalidate_host(host)
+                continue
+
+            # Now, check whether this KnownPinnedHost applies to this
+            # connection. It does if the domain either exactly matches the KPH
+            # domain, or if the KPH has include_subdomains set to True.
+            if domain == host.domain or host.include_subdomains:
+                return host
+
+
+# An object that holds information about a KnownPinnedHost. This object
+# basically just stores string data, so there's no real need to do anything
+# clever here.
+KnownPinnedHost = collections.namedtuple(
+    'KnownPinnedHost',
+    ['domain', 'pins', 'max_age', 'include_subdomains', 'report_uri', 'start_date']
+)
+
+
+def parse_public_key_pins(header, domain):
+    """
+    Parses a Public-Key-Pins header, returning a KnownPinnedHost. Invalid
+    headers cause exceptions to be thrown.
+
+    :param header: The PKP header value.
+    :param domain: The domain name used in the request. Must not be an IP
+        address.
+    """
+    # TODO: This method is super complicated, refactor.
```

Comment:
It almost certainly does, and depending on the order in which these patches land we should switch to using common code.